sql/core/src/test/resources/sql-tests/inputs/udf/udf-group-by.sql

```diff
@@ -0,0 +1,156 @@
+-- This test file was converted from group-by.sql.
+-- Test data.
+CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES
+(1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2), (null, 1), (3, null), (null, null)
+AS testData(a, b);
+
+-- Aggregate with empty GroupBy expressions.
+SELECT udf(a), udf(COUNT(b)) FROM testData;
+SELECT COUNT(udf(a)), udf(COUNT(b)) FROM testData;
+
+-- Aggregate with non-empty GroupBy expressions.
+SELECT udf(a), COUNT(udf(b)) FROM testData GROUP BY a;
+SELECT udf(a), udf(COUNT(b)) FROM testData GROUP BY b;
+SELECT COUNT(udf(a)), COUNT(udf(b)) FROM testData GROUP BY a;
```

Comment:
Does it work if we do `udf(a)` in `GROUP BY`? Seems this case is missing.

Reply:
will check.

Reply:
@HyukjinKwon I tried this one from group-analytics.sql:
```
>>> from pyspark.sql.functions import pandas_udf, PandasUDFType
>>> @pandas_udf("string", PandasUDFType.SCALAR)
... def noop(x):
...     return str(x)
... 
>>> spark.udf.register("udf", noop)
<function noop at 0x7f108c570b90>
>>> 
>>> spark.sql("CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES (1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2) AS testData(a, b)")

DataFrame[]
>>> 
>>> spark.sql("SELECT COUNT(udf(a)), COUNT(udf(b)) FROM testData GROUP BY udf(a)").show()
19/07/12 01:47:48 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
[Stage 0:>                                                          (0 + 6) / 6]19/07/12 01:47:50 ERROR Executor: Exception in task 5.0 in stage 0.0 (TID 5)
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/worker.py", line 513, in main
    process()
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/worker.py", line 505, in process
    serializer.dump_stream(out_iter, outfile)
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/serializers.py", line 398, in dump_stream
    return ArrowStreamSerializer.dump_stream(self, init_stream_yield_batches(), stream)
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/serializers.py", line 234, in dump_stream
    for batch in iterator:
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/serializers.py", line 391, in init_stream_yield_batches
    for series in iterator:
  File "<string>", line 1, in <lambda>
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/worker.py", line 106, in <lambda>
    verify_result_type(f(*a)), len(a[0])), arrow_return_type)
  File "/home/stavros/Desktop/workspace/Lightbend/spark/spark-3.0.0-SNAPSHOT-bin-lightbend/python/lib/pyspark.zip/pyspark/worker.py", line 102, in verify_result_length
    "expected %d, got %d" % (length, len(result)))
RuntimeError: Result vector from pandas_udf was not the required length: expected 1, got 29

	at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:465)
``` Is this a bug? udfs dont work with group by...

Reply:
It does:
```
>>> from pyspark.sql.functions import pandas_udf, PandasUDFType
>>> @pandas_udf("string", PandasUDFType.SCALAR)
... def noop(x):
...     return x.apply(str)
... 
>>> spark.udf.register("udf", noop)
<function noop at 0x7f2cf7d91b90>
>>> 
>>> spark.sql("CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES (1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2) AS testData(a, b)")

DataFrame[]
>>> 
>>> spark.sql("SELECT COUNT(udf(a)), COUNT(udf(b)) FROM testData GROUP BY udf(a)").show()

19/07/12 01:55:40 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
+-------------+-------------+                                                   
|count(udf(a))|count(udf(b))|
+-------------+-------------+
|            2|            2|
|            2|            2|
|            2|            2|
+-------------+-------------+

>>> 
```

Reply:
I will cover it.

Reply:
This one will not work:
```
spark.sql("SELECT udf(a + b), b, udf(SUM(a - b)) FROM testData GROUP BY udf(a + b), b WITH CUBE").show()
pyspark.sql.utils.AnalysisException: u"expression 'testdata.`a`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;;\nAggregate [udf((a#0 + b#1))#152, b#153, spark_grouping_id#149], [udf((a#0 + b#153)) AS udf((a + b))#146, b#153, udf(sum(cast((a#0 - b#1) as bigint))) AS udf(sum(cast((a - b) as bigint)))#148]\n+- Expand [List(a#0, b#1, udf((a#0 + b#1))#150, b#151, 0), List(a#0, b#1, udf((a#0 + b#1))#150, null, 1), List(a#0, b#1, null, b#151, 2), List(a#0, b#1, null, null, 3)], [a#0, b#1, udf((a#0 + b#1))#152, b#153, spark_grouping_id#149]\n   +- Project [a#0, b#1, udf((a#0 + b#1)) AS udf((a#0 + b#1))#150, b#1 AS b#151]\n      +- SubqueryAlias `testdata`\n         +- Project [a#0, b#1]\n            +- SubqueryAlias `testData`\n               +- LocalRelation [a#0, b#1]\n"

```